### ISSUE TITLE
PLIP: Retina image scales

### ISSUE THREAD

**Issue by didrix** [plone/Products.CMFPlone#1483]
**Proposer** : Diederik Veeze

**Seconder** : Maurits van Rees
# Abstract

Scale images in plone.scale such that they will look good on devices with pixel ratios higher than 1.
# Motivation

More and more devices these days can display images with higher resolution than the html/css resolution. Plone however often serves images that are scaled to the exact same size as html/css resolution, while a lot of devices can display twice the resolution.
# Assumptions

N/A
# Proposal & Implementation

I propose to add an option to the imaging control panel to set a pixel ratio for image scaling. Images generated in plone.scale will generate images whose height and width are multiplied by this pixel ratio. For instance an image in the site of 320x240 with a scaling factor of 2 should result in something like the following:
`<img src="someimage-640x480.jpeg" alt="Some Image" width="320" height="240">`
# Deliverables
- Scaling factor in plone.scale
- Option for pixel ratio in imaging control panel
- Documentation
# Risks

The image scales will get larger, but the image quality for these larger scales can be reduced while still giving a satisfactory result, for only a small increase in byte size.
# Participants

@mauritsvanrees 
# Note

~~I recently signed the contributor assignment agreement and am awaiting confirmation.~~

**Comment by ebrehault** [comment]
That's interesting.
Two remarks:

- if I understand what you propose, if we enable the 2x option in the Image control panel, then all the images will be 2 times bigger to load, even for 1x devices (I agree the rendering will be correct because of the width and height attributes, but the network impact will be bad).

Don't you think we could use the HTML5 `srcset` attribute, which allows to declare several sources according the context:
```
<img src="images/logo-nope.png"
     srcset="images/logo-big.png 1x,
          images/logo-hd.png 2x"
     width="1824" height="499" alt="">
```
?
That way, any user will load the appropriate image considering her/his device, and there is no need to force a 2x scaling factor.

- your solution (but also mine) assumes the markup will contain the width and height of the image, but right now, when a user insert an image in TinyMCE, we do not have the width and height attributes on the img tag (we just get something like `data-scale="tile"`), and that's the same with many templates where images are inserted by plone (for instance in the navigation portlet, we get the news items image with just `class="image-icon"`).
So we need to change all the `img` markup generation in Plone.
It could be done if we use `tag()` instead of `tal:attributes="src blabla"` in all the considered templates (plus fixing the TinyMCE image plugin).

Note: that's exactly the same problem @Gagaro is facing in this PR: https://github.com/plone/plone.namedfile/pull/21 (which aims to use SVG in images, for a quite similar purpose as yours, considering SVG scale automatically).

**Comment by didrix** [comment]
> if we enable the 2x option in the Image control panel, then all the images will be 2 times bigger to load

In fact images will be almost 4 times bigger. You could bring jpeg quality down however at these scales without really taking a hit in terms of how the image looks, so that images won't really be _that_ much bigger.

> we could use the HTML5 srcset attribute

Indeed, this is better. No support in IE, but fewer high res screens on devices running IE, so I don't think that's a deal breaker.

**Comment by hvelarde** [comment]
yes, I prefer @ebrehault solution: HTML5 `srcset` attribute is the future.

take a look at [collective.lazysizes](https://github.com/collective/collective.lazysizes) and the [lazysizes](https://afarkas.github.io/lazysizes/) library to see how responsiveness works with that.

supporting HTML5 `srcset` attribute in the core will make that faster and easier to implement.

big +1 on this PLIP.

**Comment by ableeb** [comment]
@didrix Welcome! Your Plone Contributors Agreement has been processed. Thank you very much for your contributions.

**Comment by hvelarde** [comment]
in fact there are only two options: picturefill and [respimg](https://github.com/aFarkas/lazysizes/tree/gh-pages/plugins/respimg); lazysizes is not a polyfill.

**Comment by espenmn** [comment]
I have tried to find ways to do this in combination with hveralrde's collective.lazysizes.
My 'solution', unfortunate, requires a hack  in alone.namedfile:
https://github.com/espenmn/collective.lazysizes/blob/master/src/collective/lazysizes/overrides/scaling.py#L110

Maybe it could be possible to 'not hardcode' which values to return
When it comes to 'scales': would it be possible to make some 'extra scales' that are 'invisible to the user', and uses these for 2x etc ?

In other words: if someone adds a scale with name 'something 100x100', a scale called 'something2x 200x200' is also created, but this does not show up in TinyMCE etc.
PS: This should be in an add-on

I put up a demo site here: 
http://xweb14f.plana.dk:8050/Lazy/images/view
(look at the html code to see the <img tag 
disable js to see 'collective.lazysizes' without the high-res images loaded

**Comment by ebrehault** [comment]
The PLIP has been approved. See https://community.plone.org/t/framework-team-meeting-minutes-2016-04-12/1943

**Comment by jensens** [comment]
The overall idea is fine, the implementation details are for discussion.

We provide content image tags by

1. image tag generation (good)
2. manual image tags in templates (bad, should be replaced by a generated tag)
3. TinyMCE generated image tags (well, may need some love also).

Given we want to use generated image tags all over the place, we have a central infrastructure to do so. With modern browsers we can use ``img`` tags ``srcset`` attribute or the ``picture`` element with a fallback to provide different images. Also if we start thinking about it we should do it right from the beginning and think about responsive images as well. Good reading here is https://responsiveimages.org/

So our image tags are generated (currently using very basic string concatenation) we can simply build something to say in a template using a (hypothetic) ``imagetag`` function::

     <img tal:replace="structure python:imagetag(context, 'fieldname', 'scalename', title='foo or generated', alt=...." />

and get out of it something like:

    <img src="http://...../@@images/fieldname/scalename"
         srcset="http://...../@@images/fieldname/scalename 1x; 
                 http://...../@@images/fieldname/scalename/@2 2x" 
         title='foo or generated' 
         alt=.......

or even generate ``<picture> <source> ....``. We may want to make this configurable, at least in a second step.

The basic two ideas behind this are:

- in the page template we do not need to know about retina, responsiveness (not covered here, but we could use the same mechanism to support it in future).
- provide another traversal element in path at the end to provide the factor for the scale
- provide a simple (some kind of template based) way to produce the tag.

Other (out of scope here, but in future) ideas: Let the image tag generator support SVG and inline it this way. The tag needed for this to show a svg could then be generated easily.

**Comment by thet** [comment]
What I found very interesting is this example from [SelfHTML]( https://wiki.selfhtml.org/wiki/HTML/Multimedia_und_Grafiken/picture ):

```
 <picture>
  <source srcset="feuerwehr-1600.jpg, feuerwehr-1600hd.jpg 2x" media="(min-width: 1024px)">
  <source srcset="feuerwehr-480.jpg, feuerwehr-480hd.jpg 2x" media="(min-width: 480px)">
  <!---Fallback--->
  <img src="feuerwehr-480.jpg" srcset="feuerwehr-320.jpg, feuerwehr-320hd.jpg 2x" alt="Feuerwehrfest 2014">
</picture>
```
This provides responsive images AND retina image scales for each of them.

**Comment by jensens** [comment]
@thet sure, this is possible. But i think this would enhance the scope of this PLIP a lot. I'd like first implement ``img`` tag variant with ``srcset`` and then its much easier to enhance this again to support responsive images. With responsive images you need also different scales and maybe different croppings (think plone.app.imagecropping) for different device width. This is also very design dependent and so difficult to implement in a generic way. So this may need an own ``picturetag(...)`` function with a bunch of parameters for the pagetemplates. And its also difficult to get this generic into TinyMCE. This would involce at least a configuration option to define which sources are used. 

So I'd say: first retina, then look further, otherwise this one gets too complex.

**Comment by mauritsvanrees** [comment]
Thanks for the comments, interesting.
Of course with one new `imagetag` standard function to rule them all, the obligatory reference is xkcd...: http://xkcd.com/927/

It *is* tempting to create such a thing though. For one thing, it makes it easier to notice if a template is still using an old way of showing images. And if needed, this one `imagetag` function (probably a browser view) could transparantly support multiple image sources, like with currently have with Archetypes and Dexterity, both needing their own kinds of tags. But Archetypes can probably be forgotten at this point.

It might be that `@@images` can simply be improved instead.

**Comment by jensens** [comment]
In fact we have this one-for-all with the ``api.get_view(context, request, "@@images").tag()`` already, but in a template its a but clumsy to use.

**Comment by mauritsvanrees** [comment]
Well, that is just `context/@@images/tag`.

**Comment by jensens** [comment]
well, for the default tag of a scale it is sufficient, if title, alt, or something else need to be passed, then its clumsy :)

**Comment by mauritsvanrees** [comment]
Sure, but then you would just need to use this:

```
<img tal:define="images context/@@images;
     tal:replace="python:images.tag(...)" />
```

With an `imagetag` function it would not really be that different, probably something like this:

```
<img tal:define="imagetag context/@@plone/imagetag;
     tal:replace="python:imagetag(...)" />
```

It does not seem to matter, except that with an `imagetag` function like you have in mind, you can pass a different `context`, which could be nice.

**Comment by jensens** [comment]
You're right, our current one is not that bad. Overall it is an implementation detail and does not matter so much.

**Comment by ebrehault** [comment]
And regarding TinyMCE, I guess we have 2 solutions:
- change the image plugin so it produces the expected markup (problem: we will have 2 versions of the same thing, 1 in Python with image.tag(), and one in JS in the plugin, so it is difficult to make it configurable)
- implement an endpoint able to receive the image url + scale and to return the proper markup, and the TinyMCE image plugin would make an AJAX call to this endpoint everytime we modify/add an image.

**Comment by jensens** [comment]
regarding TinyMCE (and overall in theory beyond) it could also be an option to use transformchain to replace specific image tags, i.e. marked with a specific class or a data attribute with an other tag.

I am just not sure if this is too in-transparent.

**Comment by ebrehault** [comment]
Using transformchain  would imply to ask the server to render+reload the currently edited TinyMCE content everytime an image id added/modified. Sounds more difficult.

**Comment by didrix** [comment]
For a first (crude) try I made some adaptations in plone.namedfile.scaling. The tag() function now generates a srcset attribute with the links to the appropriately scaled images. This seemed like the simplest solution and it seems the work quite well.

### Todo
- Now using hard coded scales of 2x and 3x which should be made configurable via the image handling settings. 
- The scales are generated regardless of whether the original file is large enough.
- Not all images are generated using the tag() function. For instance pat-plone-modal (mockup/patterns/model/pattern.js) or the icon images in the news portlet (plone.app.portlets).

## Problems
Here are some hurdles I have found:

### Waking objects
There is the problem of having to wake objects in order to generate tags from them. For instance in the news portlet images (plone.app.portlets) the image src's are generated using ```obj.getURL()+'/@@images/image/icon'``` to avoid waking the object (```obj``` is actually a brain here).

### TinyMCE
Tiny always scales the image as scale=large irrespective of how the user scales the image by dragging the corners. Dragging the corners only affect the width and height attributes of the img tag and not the image itself.

**Comment by ebrehault** [comment]
@didrix, it sounds as a good start!
Could you please provide a link to the branch where you made your changes?

**Comment by hvelarde** [comment]
awesome! I would leave optimizations to the end: it would be nice to compare computing time on before and after.

**Comment by mauritsvanrees** [comment]
Plip config is at https://github.com/plone/buildout.coredev/blob/5.1/plips/plip-1483-retina-image-scales.cfg

@didrix Note this recently merged pull request that may influence your work: https://github.com/plone/plone.namedfile/pull/24
You may want to rebase your branch on the new master.

**Comment by gforcada** [comment]
@didrix @mauritsvanrees is a PLIP jenkins job needed?

**Comment by mauritsvanrees** [comment]
@gforcada Yes, that would be good.

**Comment by gforcada** [comment]
@didrix  @mauritsvanrees the job is there: http://jenkins.plone.org/view/PLIPs/job/plip-1483-retina-image-scales/

**Comment by didrix** [comment]
@ebrehault Sorry, haven't checked github in a while.
The plipfile has been updated to include all te work I've done so far. Build with `bin/buildout -c plips/plip-1484-retina-image-scales.cfg`

Since my last comment I have added some more stuff, mainly the ability to set the scales and quality in the image handling settings. The settings page does rely on some javascript (to only show the relevant image quality fields), so you will need to rebuild the 'plone' bundle in the resource registries to get the full effect.

**Comment by hvelarde** [comment]
can we maintain this work as much version agnostic as possible? I don't see any reason this can't be backported to Plone 4.x.

comments?

**Comment by didrix** [comment]
@hvelarde There have been quite a few changes in plone.namedfile.scaling, so it won't be truly version agnostic. But you are right, it could technically be backported without too much headache. I don't know if it is desirable as plone needs to move forward with plone 5 and not stay stuck in the past by adding features to older versions.

**Comment by jensens** [comment]
This can be an addon in Plone 4.3 or could go into a 4.4 (which we do not plan). But to keep on track and from semantic versioning and imo this needs to go as a feature in 5.1.

Sorry for my plone.scale/plone.namedfile changes, but this solves a conceptual problem to a certain degree. If you need help/explanations dont hesitate to ask me.

**Comment by hvelarde** [comment]
@didrix just remember that Plone 4.3 is LTS and will be here for many years and will be maintained until Plone 6 is out. I still see no reason to migrate any of my sites to Plone 5, and this feature is no exception.

@jensens an add-on is indeed a great idea.

**Comment by ebrehault** [comment]
@didrix your code looks good to me.
As you are working on a lot of different repositories, if your implementation takes too long, you might have to rebase from master a lot and it can be sometimes painful.
I guess the TinyMCE image plugin is probably the most tricky part, and can take longer.
Maybe we should consider merging what you have done so far quite soon, and then you can focus on Tinymce image plugin.
If you are ok with this approach, try to finalize your changes (add tests for instance), and make your PRs.

**Comment by thet** [comment]
... after the pull-requests are submitted, we can do a FWT review. I'm +1 on having this finished quite soon and +1 to name the reviewers at the next @plone/framework-team meeting. I could be one of the reviewers.

**Comment by didrix** [comment]
@ebrehault Just created the first pull requests:
https://github.com/plone/Products.CMFPlone/pull/1611
https://github.com/plone/plone.namedfile/pull/25

**Comment by jensens** [comment]
@didrix What is the state of this PLIP? Whats done, whats missing, where is help needed?

**Comment by didrix** [comment]
@jensens About ready. There is still 1 robottest that fails (since the update of firefox in jenkins).

**Comment by didrix** [comment]
@ebrehault @jensens @mauritsvanrees . I believe this plip is ready to be merged. I just need to rebase plone.app.upgrade (and Products.CMFPlone I think), but don't want to do it until there is an indication that we can merge soon, otherwise I just stay busy rebasing.

**Comment by mauritsvanrees** [comment]
I did some more rebases or merges and fixes. It's being tested again in the plip job:
http://jenkins.plone.org/view/PLIPs/job/plip-1483-retina-image-scales/21/

Note: @didrix does not have much time to work on this anymore, as he needs to focus on studies, but I am a colleague and can do some of the work if needed.

**Comment by mauritsvanrees** [comment]
The PLIP job is green again. :-)

**Comment by mauritsvanrees** [comment]
Any update? Do some @plone/framework-team members still need to vote?

**Comment by jensens** [comment]
This PLIP is in state "Ready for Review". @thet and I are put ourself in the role of a reviewer, but both had not enough time. Sorry for that! I put this on top of my review list.

**Comment by mauritsvanrees** [comment]
No problem. Everyone is busy. I am just curious. Thanks!

**Comment by mauritsvanrees** [comment]
Most branches were in need of a rebase, so I did that. I did some minor fixes as well.
I added [upgrade documentation](https://github.com/plone/documentation/pull/781) too.

I would very much like to land this in the next 5.1 alpha or beta. Final reviews are very welcome.

**Comment by mauritsvanrees** [comment]
The PLIP job is successful: http://jenkins.plone.org/view/PLIPs/job/plip-1483-retina-image-scales/

(Well, the past two jobs were successful, the current one is still running, but that one should be fine too.)

**Comment by mauritsvanrees** [comment]
I noticed that we were not using the new `portal/@@image_scale` in a few places, but were still using `getObject` and then calling `@@images` on that. I have updated the `plone.app.portlets` and `plone.app.contenttypes` branches to fix that.

**Comment by thet** [comment]
Just to let you know - I reviewed all but plone.namedfile and it looks pretty good to me (so long quite straight forward...). A nice side effect is, that all image tag replacements are more concise. I'll finish the review ASAP...

**Comment by jensens** [comment]
TODO List Merge

- [x] review by @thet 
- [ ] review by @jensens  
- [ ] discuss at FTW meeting and declare as ready
- [ ] check if rebases are needed meanwhile and run plip-jon tests
- merges needed for, but from independent from PLIP (fixes/enhancements)
    - [x] plone/plone.app.caching#25
    - [x] plone/plone.app.linkintegrity#39
    - [x] plone/mockup#679
- merge together (core PLIP feature)
    - [ ] plone/Products.CMFPlone#1611 
    - [ ] plone/plone.namedfile#25 
    - [ ] plone/plone.app.upgrade#76 
    - [ ] plone/plone.app.portlets#72 
    - [ ] plone/plone.app.contenttypes#350 
    - [ ] plone/documentation#781 

@mauritsvanrees Anything's missing?

**Comment by jensens** [comment]
After Plone 5.1.b2 was released (even if still pending) ``plone.app.upgrade`` needs minor work. The current upgrade step is ``to_beta1`` and need to be moved to ``to_beta3`` (or ``to_rc1``?).

**Comment by mauritsvanrees** [comment]
The TODO list contains everything as far as I see.
I have updated `plone.app.upgrade`, and have increased the `metadata.xml` version in `CMFPlone`.
I have started the Jenkins PLIP job.

**Comment by mauritsvanrees** [comment]
I have rebased `CMFPlone` and `p.a.upgrade` and started the Jenkins job again.

**Comment by mauritsvanrees** [comment]
There were a few failing tests, but I fixed them and the plip job is green now.

**Comment by thet** [comment]
Here are some screenshots showing the improvement with Retina image scales on and off:

News Item retina off:
![screen shot 2017-03-16 at 12 57 40](https://cloud.githubusercontent.com/assets/170891/23995149/96baa37a-0a48-11e7-852a-6a2f840e94eb.png)

News Item retina on:
![screen shot 2017-03-16 at 12 57 19](https://cloud.githubusercontent.com/assets/170891/23995151/9f83f2ea-0a48-11e7-8840-c44d6b8ca8b5.png)

Image retina off:
![screen shot 2017-03-16 at 12 56 08](https://cloud.githubusercontent.com/assets/170891/23995164/b244c3be-0a48-11e7-82f2-931289ca9bed.png)

Image retina on:
![screen shot 2017-03-16 at 12 57 02](https://cloud.githubusercontent.com/assets/170891/23995227/f55a5420-0a48-11e7-8d54-0a7c85cabef0.png)

**Comment by thet** [comment]
More test results - these are the images including their size:

_Inline news image 1x, 8,7 kB:_
![n1-9ff7a15f-9b29-4322-8369-970a085369fc](https://cloud.githubusercontent.com/assets/170891/23995777/441044f6-0a4b-11e7-88c4-8af010eaf1a8.jpeg)

_Inline news image 2x, 17,5 kB:_
![n2-22165690-5d96-4178-b169-af352871cde3](https://cloud.githubusercontent.com/assets/170891/23995780/48ca1aa8-0a4b-11e7-9c35-a9a3234fbff7.jpeg)

_Inline news image 3x, 31,8 kB:_
![n3-ca6e1114-97e4-4b0a-b7d2-5232c2619009](https://cloud.githubusercontent.com/assets/170891/23995784/4dd09680-0a4b-11e7-96b1-2bb5b1e52853.jpeg)

_Image 1x, 118,3 kB:_
![img1-3df51e25-73b3-47d4-8484-e3c8025a0f55](https://cloud.githubusercontent.com/assets/170891/23995790/5414adce-0a4b-11e7-9b1a-e187c8fb4887.jpeg)

_Image 2x, 245,9 kB:_
![img2-891ced86-5bc6-4e52-954d-1a5e59f4ca17](https://cloud.githubusercontent.com/assets/170891/23995796/5aaf6142-0a4b-11e7-87bf-1cad0047f44d.jpeg)

_Image 3x, 455,5 kB:_
![img3-52d4f3da-14c9-489f-9bcc-8fb802b88747](https://cloud.githubusercontent.com/assets/170891/23995810/60d52872-0a4b-11e7-818f-ee851b377316.jpeg)

_Original Image, 5.5 MB:_
![orig-img_20161212_092308527](https://cloud.githubusercontent.com/assets/170891/23995815/66cc75b4-0a4b-11e7-9ea5-3e27c9b02d01.jpg)

**Comment by thet** [comment]
@mauritsvanrees I have added a review here: https://github.com/plone/buildout.coredev/blob/5.1/plips/reviews/plip1483-retina-image-scales-review-thet.rst

**Comment by thet** [comment]
By the way: The pics above show the "Which Hat Tower" in Minneapolis. Beautiful neighbourhood.

**Comment by hvelarde** [comment]
awesome, guys! I can't wait to test lazysizes with this new feature.

**Comment by jensens** [comment]
My review is ready and available at https://github.com/plone/buildout.coredev/blob/5.1/plips/reviews/plip1483-retina-image-scales-review-jensens.rst

**Comment by mauritsvanrees** [comment]
@jensens Did you commit and push your changes to the review? The last change to that file was two weeks ago.

**Comment by jensens** [comment]
ops, need to read more careful the console output, now its on GitHub

**Comment by thet** [comment]
@mauritsvanrees I just updated my review with a code-review section about adapterization of ``plone.namedfile.scaling.ImageScale``. This is a non-blocker, but I want to raise awareness (and not loose my thoughts while reviewing) about an issue I might need an answer to sometime soon.

Second thing, re-reading this thread, I found that you already mentioned an ``imagetag`` view with some critical thoughts about it. I mentioned the same in my review. Funny thing, I thought I came up with that idea while reviewing your code :)

**Comment by mauritsvanrees** [comment]
As for testing in the wild: you can see the `srcset` attributes in use on the carousel of a client of Zest: [Zeelandia International](http://www.zeelandia-international.com).

**Comment by jensens** [comment]
merged!

**Comment by jensens** [comment]
Thanks for the great work!

**Comment by fgrcon** [comment]
Phantastic  /  I am merging with PLIP https://github.com/plone/Products.CMFPlone/issues/1734 right now. I found still some locations to upgrade:
* plone.app.contenttypes.image.pt,
*  ....... fullview.pt  

furthermore collection portlet , event portlet. These I will fix myself during the merge.

**Comment by mauritsvanrees** [comment]
Thanks for the reviews and the merge.
Glad to have this finally in core. Thanks to colleague @didrix who did most of the work and figuring things out. Congrats on the merge!

So, reading the reviews and final comments, it seems there may be room for some extra fixes or ideas:

- [ ] An `@@imagetag` view which takes a brain as context. But I don't see much that we gain from this. A view on a brain can work fine as far as I know, but I fear a small performance dip.
- [ ] Move control panel javascript to a different place.  Currently at  [CMFPlone/static/plone.js](https://github.com/plone/Products.CMFPlone/blob/1ba0233f38a79a2ca6c5c741c855090314b06d34/Products/CMFPlone/static/plone.js#L69). But adding things to a `plone-legacy` bundle sounds strange (it's legacy, so don't add anything), and making a pattern for something that is only needed in one control panel sounds strange to me as well.
- [ ] See if we can let images added via TinyMCE use the retina stuff too.
- [ ] Maybe register `plone.namedfile.scaling.ImageScale` as browser page in `zcml` so integrators can more easily override it.
- [ ] User documentation is missing. There should at least a small chapter about how to activate and configure retina images.
- [X] Update [`plone.app.contenttypes.image.pt`](https://github.com/plone/plone.app.contenttypes/blob/master/plone/app/contenttypes/browser/templates/image.pt#L16). This already uses the `tag` function, so seems good already.
- [ ] Update [collection portlet] (https://github.com/plone/plone.portlet.collection/blob/master/plone/portlet/collection/collection.pt#L32-L36)
- [ ] Update [event portlet](https://github.com/plone/plone.app.event/blob/master/plone/app/event/portlets/portlet_events.pt#L31-L35).
- [ ] Maybe replace 'Retina' by 'HiDPI' or something like that in documentation or help texts. See https://github.com/plone/documentation/issues/812

**Comment by fgrcon** [comment]
> Update plone.app.contenttypes.image.pt. This already uses the tag function, so seems good already.

sorry I uploaded too small images, works fine!"

- [x] Update [collection portlet]  done  in branch  plip1734thumbs  (https://github.com/plone/Products.CMFPlone/issues/1734) (https://github.com/plone/plone.portlet.collection/blob/4e23864c9703b7f51b010c5a2d97de8f98455910/plone/portlet/collection/collection.pt#L19-L44)
- [x] Update [event portlet] done  in branch  plip1734thumbs (https://github.com/plone/plone.app.event/blob/6a8ebc01da833de8b32f0249204143a4ce2ad58e/plone/app/event/portlets/portlet_events.pt#L18-L35).
